Doc update (#123)
ROCmSoftwarePlatform/TransferBench: docs/how to/use-transferbench.rst

Each line in the configuration file defines a set of transfers, also known as a test, to run in parallel.

There are two ways to specify a test:

- Basic

   The basic specification assumes the same number of SEs used per transfer.
   A positive number of transfers is specified, followed by the number of SEs and triplets describing each transfer:

   Transfers SEs (srcMem1->Executor1->dstMem1) ... (srcMemL->ExecutorL->dstMemL)

   .. code-block:: bash

      1 4 (G0->G0->G1)                   Uses 4 CUs on GPU0 to copy from GPU0 to GPU1
      1 4 (G2->C1->G0)                   Uses 4 CUs on GPU2 to copy from CPU1 to GPU0
      2 4 G0->G0->G1 G1->G1->G0          Copies from GPU0 to GPU1, and GPU1 to GPU0, each with 4 SEs

- Advanced

   A negative number of transfers is specified, followed by quintuplets describing each transfer.
   A non-zero number of bytes specified will override any provided value.

   Transfers (srcMem1->Executor1->dstMem1 SEs1 Bytes1) ... (srcMemL->ExecutorL->dstMemL SEsL BytesL)

   .. code-block:: bash

      -2 (G0 G0 G1 4 1M) (G1 G1 G0 2 2M) Copies 1Mb from GPU0 to GPU1 with 4 SEs and 2Mb from GPU1 to GPU0 with 2 SEs

> Here is the list of arguments:

Author: @gilbertlee-amd Where are these arguments specified? In config file?

Reviewer: https://github.com/ROCm/TransferBench/blob/a7315688c50d84af9cf1f9f100bc0b12d51f0b4c/examples/example.cfg#L32

Yes

Author: @gilbertlee-amd Can we show example demonstrating use or arguments in the config file? The examples shown such as:  1 4 (G0->G0->G1)  don't show the use of arguments in the config file.

       | - D: DMA-executor  (indexed from 0 to GPUs - 1)

   * - dstMemL
     - Destination memory locations (where the data is written)

   * - bytesL
     - | Number of bytes to copy (use command-line specified size when 0)
       | Must be a multiple of 4 and may be suffixed with ('K','M', or 'G')
       | Memory locations are specified by one or more device characters or device index pairs.
       | Characters indicate memory type and are followed by device index (0-indexed)
       | Here are the characters and their respective memory locations:
       | - C:    Pinned host memory       (on NUMA node, indexed from 0 to [NUMA nodes-1])
       | - U:    Unpinned host memory     (on NUMA node, indexed from 0 to [NUMA nodes-1])
       | - B:    Fine-grain host memory   (on NUMA node, indexed from 0 to [NUMA nodes-1])
       | - G:    Global device memory     (on GPU device, indexed from 0 to [GPUs - 1])
       | - F:    Fine-grain device memory (on GPU device, indexed from 0 to [GPUs - 1])
       | - N:    Null memory              (index ignored)

Round brackets and arrows "->" can be included for human clarity, but will be ignored.

> Lines starting with will be ignored. Lines starting with will be echoed to output

Author: @gilbertlee-amd Lines starting with what?

Reviewer: Should be '#' for ignored lines
and '##' for echoed lines

Reviewer: This is word-for-word with here:
https://github.com/ROCm/TransferBench/blob/a7315688c50d84af9cf1f9f100bc0b12d51f0b4c/examples/example.cfg#L61

Author: Done.

.. meta::
  :description: TransferBench is a utility to benchmark simultaneous transfers between user-specified devices (CPUs or GPUs)
  :keywords: Using TransferBench, TransferBench Usage, TransferBench How To, API, ROCm, documentation, HIP

.. _using-transferbench:

---------------------
Using TransferBench
---------------------

> You can control the SRC and DST memory locations by indicating the memory type followed by the device index. TransferBench supports the following memories:

Reviewer 2: ```suggestion
You can control the SRC and DST memory locations by indicating the memory type followed by the device index. TransferBench supports the following memory types:
```

> * Coarse-grained pinned host memory

Reviewer 2: you can remove the term "memory" from this bulleted list. Redundant with the intro in play

Author: Done

* Unpinned host memory
* Fine-grained host memory
* Coarse-grained global device memory
* Fine-grained global device memory
* Null memory (for an empty transfer)

In addition, you can determine the size of the transfer (number of bytes to copy) for the tests.

> You can also specify transfer executors . The options are CPU, kernel-based GPU, and SDMA-based GPU (DMA) executors. TransferBench also provides the option to choose the number of Sub-Executors (SE). The number of SEs specifies the number of CPU threads in the case of a CPU executor and the number of compute units (CU) for a GPU executor.

Reviewer 2: ```suggestion
You can also specify transfer executors. The options are CPU, kernel-based GPU, and SDMA-based GPU (DMA) executors. TransferBench also provides the option to choose the number of Sub-Executors (SE). The number of SEs specifies the number of CPU threads in the case of a CPU executor and the number of compute units (CU) for a GPU executor.
```

.. meta::

>   :description: TransferBench documentation

Reviewer 2: Improve this

Author: Done.

>   :keywords: TransferBench, API, ROCm, documentation, HIP

Reviewer 2: This looks like a repeat of other topics in here. 
Add relevant keywords.

Author: Done

.. _using-transferbench:

---------------------
Using TransferBench
---------------------

You can control the SRC and DST memory locations by indicating the memory type followed by the device index. TransferBench supports the following memories:

* Coarse-grained pinned host memory
* Unpinned host memory
* Fine-grained host memory
* Coarse-grained global device memory
* Fine-grained global device memory
* Null memory (for an empty transfer)

In addition, you can determine the size of the transfer (number of bytes to copy) for the tests.

> You can also specify executors of the transfer. The options are CPU, kernel-based GPU, and SDMA-based GPU (DMA) executors. TransferBench also provides the option to choose the number of Sub-Executors (SE). The number of SEs specify the number of CPU threads in case of a CPU executor and the number of compute units (CU) for a GPU executor.
> For a DMA executor, the SE argument determines the number of streams to be used.

Reviewer 2: ```suggestion
You can also specify transfer executors . The options are CPU, kernel-based GPU, and SDMA-based GPU (DMA) executors. TransferBench also provides the option to choose the number of Sub-Executors (SE). The number of SEs specifies the number of CPU threads in the case of a CPU executor and the number of compute units (CU) for a GPU executor.
For a DMA executor, the SE argument determines the number of streams to be used.
```

You can specify the transfers in a configuration file or use preset configurations for transfers.

Specifying transfers in a configuration file
----------------------------------------------

> A transfer is defined as a single operation where an executor reads and adds together values from SRC memory locations, then writes the sum to the DST memory locations.

Reviewer 2: This sentence is awkward. The "then" is messed up. Should this be two sentences or reworded.

Author: Done. "A transfer is defined as a single operation where an executor reads and adds together values from SRC memory locations followed by writing the sum to the DST memory locations."

> This simplifies to a simple copy operation when dealing with a single SRC or DST.

Reviewer 2: ```suggestion
This simplifies to a copy operation when using a single SRC or DST.
```

> Here is a copy operation from a single SRC to DST:

Reviewer 2: ```suggestion
Here's a copy operation from a single SRC to DST:
```

> Three Executors are supported by TransferBench::

Reviewer 2: Is Executors supposed to have initial cap?

Author: Corrected.

> Executor:        SubExecutor:

Reviewer 2: The formatting looks messed up:
![image](https://github.com/user-attachments/assets/f154021b-e002-42de-901a-9fd80b5b5a96)

Reviewer 2: Should they be single lines

Author: Corrected.

1. CPU           CPU thread
2. GPU           GPU threadblock/Compute Unit (CU)
3. DMA           N/A (Can only be used for a single SRC to DST copy)

Each line in the configuration file defines a set of transfers, also known as a test, to run in parallel.

There are two ways to specify a test:

- **Basic**

  The basic specification assumes the same number of SEs used per transfer.
  A positive number of transfers is specified, followed by the number of SEs and triplets describing each transfer:

  Transfers SEs (srcMem1->Executor1->dstMem1) ... (srcMemL->ExecutorL->dstMemL)

  The arguments used to specify transfers in the config file are described in the :ref:`arguments table <config_file_arguments_table>`.

  **Example**:

  .. code-block:: bash

   1 4 (G0->G0->G1)                   Uses 4 CUs on GPU0 to copy from GPU0 to GPU1
   1 4 (G2->C1->G0)                   Uses 4 CUs on GPU2 to copy from CPU1 to GPU0
   2 4 G0->G0->G1 G1->G1->G0          Copies from GPU0 to GPU1, and GPU1 to GPU0, each with 4 SEs

- **Advanced**

  A negative number of transfers is specified, followed by quintuplets describing each transfer.

>   A non-zero number of bytes specified, overrides any provided value.

Reviewer 2: These two lines are not sentences - something is messed up.

Author: Changed it to:
In the advanced specification, a negative number of transfers is specified, followed by quintuplets describing each transfer.
Specifying a non-zero number of bytes overrides any provided value.

  The arguments used to specify transfers in the config file are described in the :ref:`arguments table <config_file_arguments_table>`.

  **Example**:

  .. code-block:: bash

   -2 (G0 G0 G1 4 1M) (G1 G1 G0 2 2M) Copies 1Mb from GPU0 to GPU1 with 4 SEs and 2Mb from GPU1 to GPU0 with 2 SEs

Here is the list of arguments used to specify transfers in the config file:

.. _config_file_arguments_table:

.. list-table::
   :header-rows: 1

   * - Argument
     - Description

   * - Transfers
     - Number of transfers to be run in parallel

   * - SE
     - Number of SEs to use (CPU threads or GPU threadblocks)

   * - srcMemL
     - Source memory locations (where the data is read)

   * - Executor
     - | Executor is specified by a character indicating type, followed by the device index (0-indexed):
       | - C: CPU-executed  (indexed from 0 to NUMA nodes - 1)
       | - G: GPU-executed  (indexed from 0 to GPUs - 1)
       | - D: DMA-executor  (indexed from 0 to GPUs - 1)

   * - dstMemL
     - Destination memory locations (where the data is written)

   * - bytesL
     - | Number of bytes to copy (use command-line specified size when 0).

>        | Must be a multiple of 4 and may be suffixed with ('K','M', or 'G').

Reviewer 2: ```suggestion
       | Must be a multiple of four and can be suffixed with ('K','M', or 'G').
```

       | Memory locations are specified by one or more device characters or device index pairs.
       | Characters indicate memory type and are followed by device index (0-indexed).
       | Here are the characters and their respective memory locations:
       | - C:    Pinned host memory       (on NUMA node, indexed from 0 to [NUMA nodes-1])
       | - U:    Unpinned host memory     (on NUMA node, indexed from 0 to [NUMA nodes-1])
       | - B:    Fine-grain host memory   (on NUMA node, indexed from 0 to [NUMA nodes-1])
       | - G:    Global device memory     (on GPU device, indexed from 0 to [GPUs - 1])
       | - F:    Fine-grain device memory (on GPU device, indexed from 0 to [GPUs - 1])
       | - N:    Null memory              (index ignored)

Round brackets and arrows "->" can be included for human clarity, but will be ignored.
Lines starting with # will be ignored. Lines starting with ## will be echoed to output.

> **Transfer examples:**

Reviewer 2: There is inconsistency in the space between GPU and the number in this section. Review all of them. 
![image](https://github.com/user-attachments/assets/c41a7181-0d20-4d09-a437-b2aa946a4980)

Author: Corrected.

   1 16 (G0->G0->G0G1)

.. note::

>    Running TransferBench with no arguments displays usage instructions and detected topology information.

Author: @gilbertlee-amd Can we show how to run transferbench?

Reviewer: Not quite sure what you mean.  You mean like "./TransferBench"?

Author: > Not quite sure what you mean. You mean like "./TransferBench"?

@gilbertlee-amd Yes, showing the command with arguments.

Reviewer: What arguments though?
For a specific preset?  
Not all examples may work on different systems based on how many GPUs are on the machine.

```
./TransferBench <configFile>
./TransferBench cmdline 64M "1 8 G0 G0 G1"
./TransferBench p2p
./TransferBench healthcheck
```

Author: @gilbertlee-amd The usage of arguments with ./TransferBench command is slightly complicated and confusing. I guess I would omit this for now as it would again require to explain the arguments and usage as per the use-case.

   1 16 (G0->G0->G0G1)

.. note::

   Running TransferBench with no arguments displays usage instructions and detected topology information.

Using preset configurations
------------------------------

Here is the list of preset configurations that can be used instead of configuration files:

.. list-table::
   :header-rows: 1

   * - Configuration
     - Description

   * - ``a2a``
     - All-to-all benchmark test

   * - ``cmdline``

>      - Allows transfers to run from command-line instead of a configuration file

Reviewer 2: ```suggestion
     - Allows transfers to run from the command line instead of a configuration file
```

   * - ``healthcheck``

>      - Simple health check (supported on MI300 series only)

Reviewer 2: ```suggestion
     - Simple health check (supported on AMD Instinct MI300 series only)
```

   * - ``p2p``
     - Peer-to-peer benchmark test

   * - ``pcopy``
     - Benchmark parallel copies from a single GPU to other GPUs

   * - ``rsweep``
     - Random sweep across possible sets of transfers

   * - ``rwrite``
     - Benchmark parallel remote writes from a single GPU to other GPUs

   * - ``scaling``
     - GPU subexecutor scaling tests

   * - ``schmoo``

>      - Read or write or copy operation on local or remote between two GPUs

Reviewer 2: ```suggestion
     - Read, write, or copy operation on local or remote between two GPUs
```
